# Dieses Skript sorgt dafür, dass die CSV in die PostgreSQL Datenbank geladen wird!

In [48]:
# csv laden
import pandas as pd
df = pd.read_csv("flights_clean.csv") 

In [49]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246209 entries, 0 to 246208
Data columns (total 29 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   airline_id             246209 non-null  object 
 1   airline                246209 non-null  object 
 2   flight                 246209 non-null  int64  
 3   aircraft_id            245227 non-null  object 
 4   scheduled_departure    246209 non-null  object 
 5   departure              240374 non-null  object 
 6   departure_delay        240374 non-null  float64
 7   scheduled_arrival      246209 non-null  object 
 8   arrival                239943 non-null  object 
 9   arrival_delay          239359 non-null  float64
 10  air_time               239359 non-null  float64
 11  distance               246209 non-null  float64
 12  manufacturer           239060 non-null  object 
 13  model                  239060 non-null  object 
 14  year                   238033 non-nu

In [50]:
df.head()

airline_id                 airline  flight aircraft_id  scheduled_departure  \
0         NK        Spirit Air Lines     165      N671NK  2024-06-01 05:00:00   
1         AA  American Airlines Inc.     878      N339PL  2024-06-01 05:03:00   
2         NK        Spirit Air Lines     433      N660NK  2024-06-01 05:05:00   
3         AA  American Airlines Inc.     411      N968AN  2024-06-01 05:09:00   
4         AA  American Airlines Inc.     353      N306RC  2024-06-01 05:30:00   

             departure  departure_delay    scheduled_arrival  \
0  2024-06-01 04:59:00             -1.0  2024-06-01 08:06:00   
1  2024-06-01 04:56:00             -7.0  2024-06-01 07:04:00   
2  2024-06-01 05:00:00             -5.0  2024-06-01 07:53:00   
3                  NaN              NaN  2024-06-01 08:15:00   
4                  NaN              NaN  2024-06-01 08:29:00   

               arrival  arrival_delay  ...  \
0  2024-06-01 07:34:00          -32.0  ...   
1  2024-06-01 06:32:00          -32.0  ...   
2  2024-06-01 07:25:00          -28.0  ...   
3                  NaN            NaN  ...   
4                  NaN            NaN  ...   

                            name_origin  latitude_origin longitude_origin  \
0  Newark Liberty International Airport        40.692501       -74.168701   
1  Newark Liberty International Airport        40.692501       -74.168701   
2  Newark Liberty International Airport        40.692501       -74.168701   
3  Newark Liberty International Airport        40.692501       -74.168701   
4  John F Kennedy International Airport        40.639801       -73.778900   

  destination                                 name_destination  \
0         FLL  Fort Lauderdale Hollywood International Airport   
1         CLT          Charlotte Douglas International Airport   
2         MCO                    Orlando International Airport   
3         MIA                      Miami International Airport   
4         MIA                      Miami International Airport   

   latitude_destination  longitude_destination   weekday hour cancelled  
0             26.072599             -80.152702  Saturday    5     False  
1             35.214001             -80.943100  Saturday    5     False  
2             28.429399             -81.308998  Saturday    5     False  
3             25.793200             -80.290604  Saturday    5      True  
4             25.793200             -80.290604  Saturday    5      True  

[5 rows x 29 columns]

In [ ]:
# ich will eine eindeutige Flug_iD haben --> Ist wichtig für get und delete in der API
# neue Spalte --> kopiert flight und setzt dann index +1 ran 
df['flight_id'] = df['flight'].astype(str) + (df.index + 1).astype(str)

In [52]:
df.head()

airline_id                 airline  flight aircraft_id  scheduled_departure  \
0         NK        Spirit Air Lines     165      N671NK  2024-06-01 05:00:00   
1         AA  American Airlines Inc.     878      N339PL  2024-06-01 05:03:00   
2         NK        Spirit Air Lines     433      N660NK  2024-06-01 05:05:00   
3         AA  American Airlines Inc.     411      N968AN  2024-06-01 05:09:00   
4         AA  American Airlines Inc.     353      N306RC  2024-06-01 05:30:00   

             departure  departure_delay    scheduled_arrival  \
0  2024-06-01 04:59:00             -1.0  2024-06-01 08:06:00   
1  2024-06-01 04:56:00             -7.0  2024-06-01 07:04:00   
2  2024-06-01 05:00:00             -5.0  2024-06-01 07:53:00   
3                  NaN              NaN  2024-06-01 08:15:00   
4                  NaN              NaN  2024-06-01 08:29:00   

               arrival  arrival_delay  ...  latitude_origin  longitude_origin  \
0  2024-06-01 07:34:00          -32.0  ...        40.692501        -74.168701   
1  2024-06-01 06:32:00          -32.0  ...        40.692501        -74.168701   
2  2024-06-01 07:25:00          -28.0  ...        40.692501        -74.168701   
3                  NaN            NaN  ...        40.692501        -74.168701   
4                  NaN            NaN  ...        40.639801        -73.778900   

  destination                                 name_destination  \
0         FLL  Fort Lauderdale Hollywood International Airport   
1         CLT          Charlotte Douglas International Airport   
2         MCO                    Orlando International Airport   
3         MIA                      Miami International Airport   
4         MIA                      Miami International Airport   

   latitude_destination  longitude_destination   weekday hour cancelled  \
0             26.072599             -80.152702  Saturday    5     False   
1             35.214001             -80.943100  Saturday    5     False   
2             28.429399             -81.308998  Saturday    5     False   
3             25.793200             -80.290604  Saturday    5      True   
4             25.793200             -80.290604  Saturday    5      True   

  flight_id  
0      1651  
1      8782  
2      4333  
3      4114  
4      3535  

[5 rows x 30 columns]

In [53]:
import os
from dotenv import load_dotenv
from sqlalchemy import create_engine

# .env-Datei laden (lokal nötig, Docker macht das automatisch)
load_dotenv()

# Werte aus .env lesen (POSTGRES_USER, POSTGRES_PASSWORD)
POSTGRES_USER = os.getenv("POSTGRES_USER")
POSTGRES_PASSWORD = os.getenv("POSTGRES_PASSWORD")

# Verbindung zu PostgreSQL herstellen, user und passwort muss wie in docker sein.
# Port und Datenbankname müssen ebenfalls stimmen.
engine = create_engine(f"postgresql+psycopg2://{POSTGRES_USER}:{POSTGRES_PASSWORD}@localhost:5432/flights")

df.to_sql('flights', engine, if_exists='replace', index=False) 
# flights = Name der Zieltabelle in der Datenbank
# SQLAlchemy Engine, die die Verbindung zur DB herstellt
# 'replace' -> Daten werden gelöscht, sodass nur die hochgeladen sind, die ich mit diesem durchlauf hochgeladen habe
# index = False --> Pandas Index ist keine Spalte
                    

209

Oft wird der AusgabeWert (209) auf die Größe des letzten Chunks begrenzt??? --> es ist aufjedenfall nicht die Zeilenanzahl. 

SELECT COUNT(*) FROM flights; --> sagt nämlich korrekterweise 246209